sys/include/iec62056/21.h

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (C) 2023 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_iec62056_21 IEC62056-21
+ * @ingroup     sys_iec62056
+ * @brief       IEC62056-21 Direct local data exchange
+ *
+ * IEC 62056-21 functions for direct local data exchange of electricity metering.
+ *
+ * Function implement protocol  telegram parsing.
+ *
+ * @{
+ *
+ * @file
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ */
+
+#ifndef IEC62056_21_H
+#define IEC62056_21_H
+
+#include <time.h>
+#include "event.h"
+#include "iec62056/obis.h"
+
+/**
+ * @brief Data object representing an address and value combination from a telegram
+ *
+ * The @ref value_index indicates the index of the value when an address has multiple values
+ * associated.
+ */
+typedef struct {
+    iec62056_obis_t obis;   /**< Address of the dataset */
+    const char *value;      /**< Pointer to the raw value inside the telegram, not null terminated */
+    size_t value_length;    /**< Length in number of character of the value */
+    unsigned value_index;   /**< Index of the value for the dataset */
+} iec62056_21_dataset_t;
+
+/**
+ * @brief Check if a dataset contains a value
+ *
+ * @param   dataset     The dataset to check
+ *
+ * @returns true if the dataset contains a value
+ * @returns false if the dataset only contains an address
+ */
+static inline bool iec62056_21_dataset_without_value(const iec62056_21_dataset_t *dataset)
+{
+    return dataset->value == NULL;
+}
+
+/**
+ * @brief Checks if the value of the dataset matches register representation
+ *
+ * @note Simple integer values also match this format.
+ *
+ * @param   dataset     The dataset to check
+ * @returns true if the dataset value matches the register format
+ * @returns false if the dataset value does not match the register format
+ */
+bool iec62056_21_dataset_is_register(const iec62056_21_dataset_t *dataset);
+
+/**
+ * @brief Checks if the value of the dataset matches a timestamp
+ *
+ * @param   dataset     The dataset to check
+ * @returns true if the dataset value matches the time format
+ * @returns false if the dataset value does not match the time format
+ */
+bool iec62056_21_dataset_is_timestamp(const iec62056_21_dataset_t *dataset);
+
+/**
+ * @brief Parse the numerical part of a register value to a decimal fraction.
+ *
+ * Value is parsed to @p mantissa * 10 ^ @p scaling
+ *
+ * This function does not use floating point operations
+ *
+ * @param   dataset     The dataset to parse
+ * @param   mantissa    Resulting mantissa
+ * @param   scaling     Resulting scaling factor
+ *
+ * @returns negative when parsing failed, zero on success
+ */
+int iec62056_21_dataset_parse_register_to_decfrac(const iec62056_21_dataset_t *dataset,
+                                                  int64_t *mantissa, int64_t *scaling);
```

Comment:
Do we really get 128 bit values from this?

Reply:
From IEC62056-21, on the length of the data:

> Value: 32 printable characters maximum with the exception of (, ), *, / and !. For decimal values, only points (not commas) shall be used and shall be counted as characters.

Values can be large here, but I don't know if those appear in the wild, I'm severely lacking real world test cases to be honest.